flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherGatewayServiceFactory.java

```diff
@@ -118,7 +120,12 @@ public ApplicationDispatcherGatewayServiceFactory(
         return DefaultDispatcherGatewayService.from(dispatcher);
     }
 
-    private List<JobID> getRecoveredJobIds(final Collection<JobGraph> recoveredJobs) {
-        return recoveredJobs.stream().map(JobGraph::getJobID).collect(Collectors.toList());
+    private Set<JobID> getRecoveredJobIds(
+            final Collection<JobGraph> recoveredJobs,
+            final Collection<JobResult> recoveredDirtyJobs) {
+        return Stream.concat(
+                        recoveredJobs.stream().map(JobGraph::getJobID),
+                        recoveredDirtyJobs.stream().map(JobResult::getJobId))
+                .collect(Collectors.toSet());
```

Comment:
This part I'm not sure about. Could you document why we need to pass dirty jobs into ApplicationDispatcherBootstrap? Why is it not enough to just pass them into the dispatcher.

(I'm not saying it's incorrect, just can't reason about it from top of my head)